lib/statsig_logging.ex

```diff
@@ -0,0 +1,123 @@
+defmodule Statsig.Logging do
+  use GenServer
+  require Logger
+
+  @type state :: %{
+    flush_interval: integer() | nil,
+    events: list(),
+    flush_timer: reference() | nil
+  }
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, %{
+      flush_interval: 60_000,
+      events: [],
+      flush_timer: nil,
+    }, name: __MODULE__)
+  end
+
+  def initialize(flush_interval) do
+    GenServer.call(__MODULE__, {:initialize, flush_interval})
+  end
+
+  def shutdown do
+    GenServer.call(__MODULE__, :shutdown)
+  end
+
+  @impl true
+  def init(state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_call(:shutdown, _from, state) do
+    Logger.info("Shutting down Statsig.Logging")
+
+
+    new_state = if length(state.events) > 0 do
+      Logger.info("Flushing #{length(state.events)} events before shutdown")
+      flush_events(state)
+    else
+      state
+    end
+
+    if state.flush_timer, do: Process.cancel_timer(state.flush_timer)
+
+    new_state = %{new_state |
+      flush_timer: nil,
+      flush_interval: nil,
+    }
+
+    {:reply, :ok, new_state}
+  end
+
+
+  @impl true
+  def handle_call({:initialize, flush_interval}, _from, state) do
+    new_state = if is_number(flush_interval) do
+      Map.put(state, :flush_interval, flush_interval)
+    else
+      state
+    end
+
+    new_state = if new_state.flush_interval != nil do
+      timer = Process.send_after(__MODULE__, :flush_events, new_state.flush_interval)
+      Map.put(new_state, :flush_timer, timer)
+    else
+      new_state
+    end
+
+    {:reply, :ok, new_state}
+  end
+
+  @impl true
+  def handle_call(:flush, _from, state) do
+    new_state = flush_events(state)
+    unsent_events = new_state.events
+    {:reply, unsent_events, new_state}
+  end
+
+  @impl true
+  def handle_info(:flush_events, state) do
+    new_state = if state.flush_interval != nil do
+      timer = Process.send_after(__MODULE__, :flush_events, state.flush_interval)
+      Map.put(state, :flush_timer, timer)
+    else
+      state
```

Comment:
this code has been duplicated a couple times, make a function that does it, or better yet, move it into the state. 

```elixir 
def schedule_flush(%State{} = state) do
  case state.flush_timer do 
     ref when is_ref(ref) -> Process.cancel_timer(ref)
     _ -> :ok 
  end 

  ref = Process.send_after(self(), :flush_events, state.flush_interval) 
  %State{state | flush_timer: ref}
end
```

```elixir
new_state = state 
|> State.flush_events()
|> State.schedule_flush()
```